Add/Fix Zombienet Docs (#8)
papermoonio/polkadot-ecosystem-docs-draft: docs/dev-tools/zombienet/overview.md

> === "Using the Executable" 

Reviewer: Got the ok to keep the tabs for this Installation section. It does seem like there are a lot of `note` and `warning` sections which may result in people skipping over them, etc. My suggestion is to review what the [Google documentation guide](https://developers.google.com/style/notices) says about notices and rework as many of these as you can into regular text. Keep only the really important things as notes and warnings.

    spawn <INSERT_ZOMBIENET_CONFIG_FILE_NAME>.toml
    ```

>     To run the command above, you need to have [nix flakes enabled](https://nixos.wiki/wiki/Flakes#Enable_flakes){target=_blank}.

Reviewer 2: ```suggestion
    To run the command above, you need to have [Flakes](https://nixos.wiki/wiki/Flakes#Enable_flakes){target=_blank} enabled.
```

    npm run zombie -- -p native spawn host-current-files/minimal.toml
    ```

    The command above mounts the current directory to the `/workspace` directory inside the Docker container. This allows Zombienet to access the configuration file and other files in the current directory. If you want to mount a different directory, replace `$(pwd)` with the desired directory path.

## Providers

> Zombienet is a JavaScript library designed to run on Node.js and support different backend providers in running the nodes. At this moment, [Kubernetes](https://kubernetes.io/){target=_blank}, [Podman](https://podman.io/){target=_blank}, and local are supported.

Reviewer 2: I don't think mentioning that Zombienet is a Javascript library is necessary in this section. You could add that to the introduction instead

    npm run zombie -- -p native spawn host-current-files/minimal.toml
    ```

    The command above mounts the current directory to the `/workspace` directory inside the Docker container. This allows Zombienet to access the configuration file and other files in the current directory. If you want to mount a different directory, replace `$(pwd)` with the desired directory path.

## Providers

> Zombienet supports different backend providers in running the nodes. At this moment, [Kubernetes](https://kubernetes.io/){target=_blank}, [Podman](https://podman.io/){target=_blank}, and local are supported.

Reviewer 3: ```suggestion
Zombienet supports different backend providers for running the nodes. At this moment, [Kubernetes](https://kubernetes.io/){target=_blank}, [Podman](https://podman.io/){target=_blank}, and local are supported, which can be declared as `kubernetes`, `podman`, or `native`, respectively.
```

...
```

> At the moment, Zombienet supports the following providers: `kubernetes`, `podman`, and `native`.

Author: Yes, that makes sense. I kept using the word `native` in some cases because it is the keyword recognized by Zombienet, but adding a small clarification like those parentheses would make it clearer 🚀

It's important to note that each provider has specific requirements and associated features. The subsequent sections will guide you through the installation process for each provider and the requirements and features each provider offers.

### Kubernetes

#### Requirements

Zombienet is designed to be compatible with a variety of Kubernetes clusters, including [Google Kubernets Engine (GKE)](https://cloud.google.com/kubernetes-engine){target=_blank}, [Docker Desktop](https://docs.docker.com/desktop/kubernetes/){target=_blank}, and [kind](https://kind.sigs.k8s.io/){target=_blank}. To effectively interact with your cluster, you'll need to ensure that [`kubectl`](https://kubernetes.io/docs/reference/kubectl/) is installed on your system, which is the Kubernetes command-line tool that allows you to run commands against Kubernetes clusters.

Moreover, in order to create resources such as namespaces, pods, and cronJobs within the target cluster, you must have the appropriate permissions granted to your user or service account. These permissions are essential for managing and deploying applications effectively within Kubernetes.

#### Features

> In Kubernetes, Zombienet uses the Prometheus operator (if available) to oversee monitoring and visibility. This configuration ensures that only essential networking-related pods are deployed. Using the Prometheus operator, Zombienet improves its ability to efficiently monitor and manage network activities within the Kubernetes cluster. 

Author: For sure! Please feel free to be as meticulous as you need. It would help me to grasp the flow quicker 😎

Moreover, in order to create resources such as namespaces, pods, and cronJobs within the target cluster, you must have the appropriate permissions granted to your user or service account. These permissions are essential for managing and deploying applications effectively within Kubernetes.

#### Features

In Kubernetes, Zombienet uses the Prometheus operator (if available) to oversee monitoring and visibility. This configuration ensures that only essential networking-related pods are deployed. Using the Prometheus operator, Zombienet improves its ability to efficiently monitor and manage network activities within the Kubernetes cluster. 

### Podman

#### Requirements

Zombienet supports Podman rootless as a provider. To use Podman as a provider, you need to have Podman installed on your system. Podman is a daemonless container engine for developing, managing, and running Open Container Initiative (OCI) containers and container images on Linux-based systems. You can install Podman by following the instructions provided on the [Podman website](https://podman.io/getting-started/installation){target=_blank}.

> !!! note

Reviewer 2: Maybe change this to a warning?

    Currently, Podman can only be used with Zombienet on Linux machines. Although Podman has support for macOS through an internal VM, the Zombienet provider code requires Podman to run natively on Linux.

#### Features
    

> Using Podman, Zombienet deploys additional pods to enhance the monitoring and visibility of the active network. Specifically, pods for Prometheus, Tempo, and Grafana are included in the deployment. Grafana is configured with Prometheus and Tempo as data sources.

Reviewer 2: I would add links to Prometheus, Tempo and Grafana. The user may not know what they are

Author: Those links are added at the very beginning of this section:

> ## Providers
> 
> Zombienet supports different backend providers in running the nodes. At this moment, [Kubernetes](https://kubernetes.io/){target=_blank}, [Podman](https://podman.io/){target=_blank}, and local are supported.

Upon launching Zombienet, access to these monitoring services is facilitated through specific URLs provided in the output:

- Prometheus - [http://127.0.0.1:34123](http://127.0.0.1:34123){target=_blank}
- Tempo - [http://127.0.0.1:34125](http://127.0.0.1:34125){target=_blank}
- Grafana - [http://127.0.0.1:41461](http://127.0.0.1:41461){target=_blank}

It's important to note that Grafana is deployed with default admin access.

> !!! note
>     When network operations cease—either by halting a running spawn with Ctrl+C or upon completion of the test—Zombienet automatically removes all associated pods, including those for Prometheus, Tempo, and Grafana.

Reviewer 2: ```suggestion
!!! note
    When network operations cease—either by halting a running spawn with Ctrl+C or upon completion of the test—Zombienet automatically removes all associated pods.
```

### Local

#### Requirements
    

> The Zombienet local provider, also referred to as native, enables you to run nodes as local processes in your environments. You must have the necessary binaries for your network (such as `polkadot` and `polkadot-parachain`). These binaries should be available in your PATH, allowing Zombienet to spawn the nodes as local processes.

Reviewer 2: ```suggestion
The Zombienet local provider, also referred to as native, enables you to run nodes as local processes in your environment. You must have the necessary binaries for your network (such as `polkadot` and `polkadot-parachain`). These binaries should be available in your PATH, allowing Zombienet to spawn the nodes as local processes.
```

To install the necessary binaries, you can use the Zombienet CLI command:

```bash
zombienet setup polkadot polkadot-parachain
```

This command will download and prepare the necessary binaries for Zombienet’s use.

!!! warning 
    The `polkadot` and `polkadot-parachain` binaries releases are not compatible with macOS. As a result, macOS users will need to clone the [Polkadot repository](https://github.com/paritytech/polkadot-sdk){target=_blank}, build the Polkadot binary, and manually add it to their PATH for `polkadot` and `polkadot-parachain` to work.

If you need to use a custom binary, ensure the binary is available in your PATH. You can also specify the binary path in the network configuration file. To showcase this, this guide will use the custom [Open Zeppelin template](https://github.com/OpenZeppelin/polkadot-runtime-templates){target=_blank} as an example.

First, clone the Open Zeppelin template repository:

> ```bash
> git clone https://github.com/OpenZeppelin/polkadot-runtime-templates && cd polkadot-runtime-templates/generic-template
> ```

Reviewer 2: ```suggestion
```bash
git clone https://github.com/OpenZeppelin/polkadot-runtime-templates \
&& cd polkadot-runtime-templates/generic-template
```
```

> ```bash
> git clone https://github.com/OpenZeppelin/polkadot-runtime-templates && cd polkadot-runtime-templates/generic-template
> ```

Reviewer 2: ```suggestion
```bash
git clone https://github.com/OpenZeppelin/polkadot-runtime-templates \
&& cd polkadot-runtime-templates/generic-template
```
```

...
```

> At the moment, Zombienet supports the following providers: `kubernetes`, `podman`, and `native` (local).
> 

Reviewer 3: ```suggestion
```

This just seems repetitive to me

    !!! note
        When network operations cease—either by halting a running spawn with Ctrl+C or upon completion of the test—Zombienet automatically removes all associated pods, including those for Prometheus, Tempo, and Grafana.

### Local

> === "Requirements"

Reviewer: For this providers section, they do want it broken out into sections instead of these tabs. Please rework/reformat, commit changes, and request a new review.

|  Command  |                                            Description                                             |                                                                                                                                                      Arguments                                                                                                                                                       |
| :-------: | :------------------------------------------------------------------------------------------------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |

> |  `spawn`  |                            Spawn the network defined in the config file                            |                                                          `<networkConfig>` - a file that declares the desired network to be spawned in `toml` or `json` format. For further information, check out the [Configuration Files](#configuration-files) section                                                           |

Author: Yes, they can be interchanged. For Zombienet, spawning a network using this command is equivalent:

```bash
./zombienet-macos-arm64 -p native spawn ./testing-config-files/minimal-config.toml
```

Than using this command:

```bash
./zombienet-macos-arm64  spawn ./testing-config-files/minimal-config.toml -p native
```

|  `spawn`  |                            Spawn the network defined in the config file                            |                                                          `<networkConfig>` - a file that declares the desired network to be spawned in `toml` or `json` format. For further information, check out the [Configuration Files](#configuration-files) section                                                           |
|  `test`   |                                  Run test on the network spawned                                   |                                                                      `<testFile>` - a file that defines assertions and tests against the spawned network, using natural language expressions to evaluate metrics, logs, and built-in functions                                                                       |

> |  `setup`  |                           Set up the dev environment of Zombienet ready.                           |                                                                                          `<binaries>` - executables that will be downloaded and prepared to be used by Zombienet. Options: `polkadot`, `polkadot-parachain`                                                                                          |

Reviewer 2: The phrase "Set up the dev environment of Zombienet ready." is quiet difficult to understand. I would put something like Set up the Zombienet development environment.

Then, you can use the following flags to customize the behavior of the CLI:

|                 Argument                  |                                                Description                                                |
| :---------------------------------------: | :-------------------------------------------------------------------------------------------------------: |
|            `-p`, `--provider`             |  Override provider to use (choices: `podman`, `default`, and, `native`). By default it uses `kubernetes`  |

> |           `-d`, `--dir` <path>            | Directory path for placing the network files instead of random temp one (e.g. -d /home/user/my-zombienet) |

Reviewer 3: Closest I could get to it, but on line 273. Just need to add "The" to the beginning of the "Zombienet repository also provides a folder with examples" sentence

> 

Reviewer 2: The warning above is repeated. The same warning is displayed in the Local Provider

|            `-p`, `--provider`             | Override provider to use (choices: `podman`, `kubernetes`, and, `native`). By default it uses `kubernetes`  |
|           `-d`, `--dir` <path>            | Directory path for placing the network files instead of random temp one (e.g. `-d /home/user/my-zombienet`) |
|              `-f`, `--force`              |                                     Force override all prompt commands                                      |
|        `-l`, `--logType` <logType>        |    Type of logging on the console (choices: `table`, `text`, and, `silent`). By default it uses `table`     |
|             `-m`, `--monitor`             |                               Start as monitor, do not auto clean up network                                |
| `-c`, `--spawn-concurrency` <concurrency> |                    Number of concurrent spawning process to launch. By default it is `1`                    |
|              `-h`, `--help`               |                                          Display help for command                                           |

## Configuration Files 

> The network configuration can be given in either `json` or `toml` format. The Zombienet repository also provides a [folder with some examples](https://github.com/paritytech/zombienet/tree/main/examples){target=_blank} of configuration files that can be used as a reference.

Reviewer 3: ```suggestion
The network configuration can be given in either `json` or `toml` format. The Zombienet repository also provides a [folder with some examples](https://github.com/paritytech/zombienet/tree/main/examples){target=\_blank} of configuration files that can be used as a reference.
```

| `default_substrate_cli_args_version` |    0 \| 1 \| 2    | Set the substrate cli args version                                                                               | -                 |
|         `default_overrides`          | Array of objects  | An array of overrides to upload to the nodes                                                                     | -                 |
|         `default_resources`          |      Object       | Only available in kubernetes, represent the resources limits/reservations needed by the nodes by default         | -                 |
|     `default_prometheus_prefix`      |      String       | A parameter for customizing the metric's prefix                                                                  | `substrate`       |
|      `random_nominators_count`       | Number (optional) | If set and the stacking pallet is enabled, Zombienet will generate x nominators and inject them into the genesis | -                 |
|          `max_nominations`           |      Number       | The max number of nominations allowed by a nominator. Should match the value set in the runtime                  | `24`              |

??? Nodes

>     There is one specific key capable of receiving more subkeys: the `nodes` key. This key is used to define further parameters for the nodes. The available keys:

Reviewer 3: ```suggestion
    There is one specific key capable of receiving more subkeys: the `nodes` key. This key is used to define further parameters for the nodes. The available keys are:
```
